kong/runloop/handler.lua

```diff
@@ -206,6 +215,13 @@ local function register_events()
       log(DEBUG, "[events] Service updated, invalidating router")
       cache:invalidate("router:version")
     end
+
+    if data.operation == "create" or
+       data.operation == "update" then
+      if utils.hostname_type(data.entity.host) == "name" then
+        timer_at(0, warmup_hostname_dns_timer, data.entity.host)
```

Comment:
1. Timers are a scarce resource. We need error handling allocating here, at least to prevent administrators from not realizing they've maxed-out their number of pending timers.
2. It'd be nice to have the ability to force an `asyncQuery` from lua-resty-dns-client, so we don't have to even allocate this timer in the first place, and directly rely on lua-resty-dns-client's `asyncQuery` time. Is that possible with today's lua-resty-dns-client @Tieske?

As it stands, this can _easily_ overwhelm timer resources when performing a large number of subsequent Services creations in a short period of time (e.g. from a script or from delarative config), to the point that given the below comment (on this only warming up the local node), I'm not sure this approach is worth our time.

If we could group the DNS lookups in a single timer, we would be _much_ better off already.

Reply:
Adding to the above:

3. This also only warms up a single worker, which furthermore emphasizes that this is optimizing local/testing environments but doesn't do much to help a production deployment, as far as I can tell?
4. We may also be evicting (oldest) entries from the worker's DNS' lrucache in favor of warmup DNS lookups, but without being guaranteed that the warmed-up records will be queried before the evicted records, which somewhat deafeats the purpose of our LRU cache.

Reply:
> It'd be nice to have the ability to force an asyncQuery from lua-resty-dns-client, so we don't have to even allocate this timer in the first place, and directly rely on lua-resty-dns-client's asyncQuery time. Is that possible with today's lua-resty-dns-client @Tieske?

No. The `asyncQuery` is related to the `stale_ttl` setting. Where we return stale data, but fire a refresh query in the background.

Reply:
That was my understanding as well. Alas.

Reply:
> This also only warms up a single worker, which furthermore emphasizes that this is optimizing local/testing environments but doesn't do much to help a production deployment, as far as I can tell?

The benefits would be (if warm up is done in each worker, not just 1) for scaling a Kong cluster, the newly spun up Kong node would be faster in dealing with first requests. This might actually make a difference with "wall-of-traffic" events.

But as said; only if it is done for all the workers.

Reply:
> newly spun up Kong node would be faster in dealing with first requests

This is not what this PR does. Making fresh node's requests faster is already handled by 8884973f7d7a6059370cd0cca87de3fc3edc5fa7.

Reply:
Thx @thibaultcha I missed that one. That will deal with it indeed. Considering that, imo, this PR is not worth the additional complexity.

Reply:
Agreed. @kikito, I think this was a worthwhile experiment, but how about we close this and move on?